designs/2022-eslint-to-esm/README.md

```diff
@@ -0,0 +1,212 @@
+- Repo: eslint/eslint
+- Start Date: (2022-02-28)
+- RFC PR: https://github.com/eslint/rfcs/pull/88
+- Authors: aladdin-add(weiran.zsd@outlook.com)
+
+# convert eslint codebase to esm
+
+## Summary
+
+<!-- One-paragraph explanation of the feature. -->
+Convert the ESLint codebase to ESM and publish the ESM version.
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+The Node.js community has been written commonjs for a long time, the same goes for ESLint. But with ESM introduced in ES6 and officially supported by Node.js v12, more and more projects are moving to ESM. Not long ago, we had converted a few repos under the ESLint organization to ESM:
```

Comment:
to put this in context: "more and more" is a pretty big exaggeration; there's one particular author whose outsized package count skews this impression; if they are skipped (or treated as "one" instead of "n"), it's a vanishingly small number.

Reply:
I guess you were saying sindre; but he was maintaining a vast number of packages, and cannot be skipped( as being widely used. :)

What do you think of changing to "a growing number of..."? or any other suggestions?

Reply:
I think it is very unwise to let a single person’s actions drive change in any ecosystem, regardless of how many packages they publish.

Reply:
it depends. IMHO, moving to esm-only is dialectical - it's something long-term beneficial, while also suffering short-term pain.

Reply:
What’s the benefit?

Reply:
@ljharb intentionally spreads FUD to derail progress towards standard JavaScript modules, and doesn't argue in good faith. Of course there are benefits to using ESM. The first blindingly obvious benefit to using ESM is you can use other ESM packages as dependencies. The second blindingly obvious benefit to using ESM over CJS is your modules can work in a browser and Deno. The [Ruck](https://github.com/jaydenseric/ruck) buildless web app framework would not be possible without ESM; see it working here: https://ruck.tech .

HTTP imports provide a much saner (not to mention standards aligned) way to consume published packages, where you can shop around for a CDN instead of being locked into a monetized, centralised repository with proprietary config and CLIs (npm). Import maps provide a standard and clean way to substitute dependencies in your graph, or customise them via CDN URL query string parameters, etc. The Deno community is exploring all sorts of wonderfull ideas based on a diversity of CDNs.

There are other nice things about moving everything to ESM, such as top level await, simpler/better dev tooling, improved tree shaking for bundlers, and less of a learning curve about syntax or interoperability.

It's also a falsehood to say Sindre is the only package author moving to ESM. Lots of people are; I've been moving all of my packages (> 5 million installs per week) over to pure ESM one by one. And not because I'm making some kind of statement or trying to force a change, but because I ***need*** them to be ESM to build cool stuff.

It's also wrong to mislead that npm represents the entirety of the JS world; 100% of Deno packages are pure ESM.

Reply:
The personal attack and liberal application of thumbs down reacji notwithstanding, it was a real question, asked in good faith.

> The first blindingly obvious benefit to using ESM is you can use other ESM packages as dependencies

this is only a benefit if there’s an ESM-only package one needs that isn’t available in CJS - I’m not aware of anything where this is the case, at least for eslint’s purposes.

> modules can work in a browser and Deno

The browser doesn’t apply to eslint. Deno compiles TS on the fly; there’s no reason it couldn’t compile CJS on the fly too - but either way, eslint is an npm package, so deno isn’t really relevant - but at least that would be a reasonable benefit if it’s a maintenance burden eslint wants to take on. 

> HTTP imports provide a much saner

(first of all, “saner” is an ableist term in this context; and using it probably violates eslint’s CoC)
This is an opinion, and one that very much contradicts my experience and security expertise. This brings in all the vulnerabilities and attack surfaces and nondeterminism of DNS, to a server environment where such things are far more dangerous than in the browser. npm is the superior approach - that it’s monetized is irrelevant (except in that it’s necessary in capitalism to monetize something you want to function properly), Also, import maps *are not standard yet* - “chrome and deno ship them” don’t constitute a standard.

I appreciate the additional context about your packages; it’s good to know who needs to be added to a list of authors to avoid.

Reply:
But what's the benefits of cjs(except for compat issues)?

ESM is the "standard", and a fully different world than CJS. I don't see the benefit of splitting the community into two. If we must choose one, it's esm.

Reply:
@aladdin-add interoperability and compat are one of the biggest benefits that could possible exist.

The problem is that the *choice* of ESM is what splits the community, because you can import CJS but you can't require ESM. By choosing CJS, you support both communities, and avoid creating friction for users before node's ESM implementation itself, not to mention the overarching ecosystem, is ready.

CJS is a standard too - it's precisely as standard as ESM is there.

Reply:
the interoperability and compat issue will not a strong need if the community moving to esm. Then we will have a more united  community. :)

Reply:
I don't believe that will ever happen - but, even if it would, then the time to do this migration is *after that* - not merely because some dream of it.

CJS will be around forever.